src/NServiceBus.Core/Persistence/InMemory/Outbox/InMemoryOutboxStorage.cs

```diff
@@ -48,24 +47,21 @@ public Task SetAsDispatched(string messageId, ContextBag context)
                 return TaskEx.CompletedTask;
             }
 
-            storedMessage.TransportOperations.Clear();
-            storedMessage.Dispatched = true;
-
+            storedMessage.MarkAsDispatched();
             return TaskEx.CompletedTask;
         }
 
         public void RemoveEntriesOlderThan(DateTime dateTime)
         {
-            var entriesToRemove = storage
-                .Where(e => e.Value.Dispatched && e.Value.StoredAt < dateTime)
-                .Select(e => e.Key)
-                .ToList();
-
-            foreach (var entry in entriesToRemove)
+            foreach (var entry in storage.ToArray())
```

Comment:
dont think so since it is manipulated inside the enumeration `storage.TryRemove`